File: tests/functional/Bots/Python/Skills/CodeFirst/WaterfallSkillBot/dialogs/activity_router_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import json
from typing import Dict
from datetime import datetime

from botbuilder.core import MessageFactory, ConversationState
from botbuilder.dialogs import (
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
    DialogTurnStatus,
    ComponentDialog,
)
from botbuilder.dialogs.skills import (
    SkillDialogOptions,
    SkillDialog,
    BeginSkillDialogOptions,
)
from botbuilder.schema import Activity, ActivityTypes, InputHints
from botbuilder.integration.aiohttp.skills import SkillHttpClient
from config import DefaultConfig
from skill_conversation_id_factory import SkillConversationIdFactory
from dialogs.cards import CardDialog
from dialogs.delete import DeleteDialog
from dialogs.proactive import WaitForProactiveDialog
from dialogs.message_with_attachment import MessageWithAttachmentDialog
from dialogs.auth import AuthDialog
from dialogs.sso import SsoSkillDialog
from dialogs.file_upload import FileUploadDialog
from dialogs.update import UpdateDialog

ECHO_SKILL = "EchoSkill"


class ActivityRouterDialog(ComponentDialog):
    def __init__(
        self,
        configuration: DefaultConfig,
        conversation_state: ConversationState,
        conversation_id_factory: SkillConversationIdFactory,
        skill_client: SkillHttpClient,
        continuation_parameters_store: Dict,
    ):
        super().__init__(ActivityRouterDialog.__name__)

        self.add_dialog(CardDialog(configuration))
        self.add_dialog(MessageWithAttachmentDialog(configuration))

        self.add_dialog(
            WaitForProactiveDialog(configuration, continuation_parameters_store)
        )

        self.add_dialog(AuthDialog(configuration))
        self.add_dialog(SsoSkillDialog(configuration))
        self.add_dialog(FileUploadDialog())
        self.add_dialog(DeleteDialog())
        self.add_dialog(UpdateDialog())

        self.add_dialog(
            self.create_echo_skill_dialog(
                configuration, conversation_state, conversation_id_factory, skill_client
            )
        )
        self.add_dialog(
            WaterfallDialog(WaterfallDialog.__name__, [self.process_activity])
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    def create_echo_skill_dialog(
        self,
        configuration: DefaultConfig,
        conversation_state: ConversationState,
        conversation_id_factory: SkillConversationIdFactory,
        skill_client: SkillHttpClient,
    ) -> SkillDialog:
        if configuration.SKILL_HOST_ENDPOINT is None:
            raise Exception("SkillHostEndpoint is not in configuration")

        if configuration.ECHO_SKILL_INFO is None:
            raise Exception("EchoSkillInfo is not set in configuration")

        options = SkillDialogOptions(
            bot_id=configuration.APP_ID,
            conversation_id_factory=conversation_id_factory,
            skill_client=skill_client,
            skill_host_endpoint=configuration.SKILL_HOST_ENDPOINT,
            conversation_state=conversation_state,
            skill=configuration.ECHO_SKILL_INFO,
        )

        return SkillDialog(options, ECHO_SKILL)

    async def process_activity(self, step_context: WaterfallStepContext):
        # A skill can send trace activities, if needed.
        await step_context.context.send_activity(
            Activity(
                type=ActivityTypes.trace,
                timestamp=datetime.utcnow(),
                name="ActivityRouterDialog.process_activity()",
                label=f"Got ActivityType: {step_context.context.activity.type}",
            )
        )

        if step_context.context.activity.type == ActivityTypes.event:
            return await self.on_event_activity(step_context)

        # We didn't get an activity type we can handle.
        await step_context.context.send_activity(
            activity_or_text=f'Unrecognized ActivityType: "{step_context.context.activity.type}".',
            input_hint=InputHints.ignoring_input,
        )
        return DialogTurnResult(DialogTurnStatus.Complete)

    async def on_event_activity(self, step_context: WaterfallStepContext):
        activity = step_context.context.activity
        await step_context.context.send_activity(
            Activity(
                type=ActivityTypes.trace,
                timestamp=datetime.utcnow(),
                name="ActivityRouterDialog.on_event_activity()",
                label=f"Name: {activity.name}. Value: {json.dumps(activity.value)}",
            )
        )

        if activity.name == "Cards":
            return await step_context.begin_dialog(CardDialog.__name__)

        if activity.name == "Proactive":
            return await step_context.begin_dialog(WaitForProactiveDialog.__name__)

        if activity.name == "MessageWithAttachment":
            return await step_context.begin_dialog(MessageWithAttachmentDialog.__name__)

        if activity.name == "Auth":
            return await step_context.begin_dialog(AuthDialog.__name__)

        if activity.name == "Sso":
            return await step_context.begin_dialog(SsoSkillDialog.__name__)

        if activity.name == "FileUpload":
            return await step_context.begin_dialog(FileUploadDialog.__name__)

        if activity.name == "Echo":
            # Start the EchoSkillBot
            message_activity = MessageFactory.text("I'm the echo skill bot")
            message_activity.delivery_mode = activity.delivery_mode
            dialog = await self.find_dialog(ECHO_SKILL)
            return await step_context.begin_dialog(
                dialog.id, BeginSkillDialogOptions(activity=message_activity)
            )

        if activity.name == "Delete":
            return await step_context.begin_dialog(DeleteDialog.__name__)

        if activity.name == "Update":
            return await step_context.begin_dialog(UpdateDialog.__name__)

        # We didn't get an event name we can handle.
        await step_context.context.send_activity(
            activity_or_text=f'Unrecognized EventName: "{step_context.context.activity.name}".',
            input_hint=InputHints.ignoring_input,
        )
        return DialogTurnResult(DialogTurnStatus.Complete)

File: tests/functional/Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/ActivityRouterDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Concurrent;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Builder.TraceExtensions;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Auth;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Cards;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Delete;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.FileUpload;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.MessageWithAttachment;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Proactive;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Sso;
using Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Update;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs
{
    /// <summary>
    /// A root dialog that can route activities sent to the skill to different sub-dialogs.
    /// </summary>
    public class ActivityRouterDialog : ComponentDialog
    {
        private static readonly string _echoSkill = "EchoSkill";

        public ActivityRouterDialog(IConfiguration configuration, IHttpContextAccessor httpContextAccessor, ConversationState conversationState, SkillConversationIdFactoryBase conversationIdFactory, SkillHttpClient skillClient, ConcurrentDictionary<string, ContinuationParameters> continuationParametersStore)
            : base(nameof(ActivityRouterDialog))
        {
            AddDialog(new CardDialog(httpContextAccessor));
            AddDialog(new MessageWithAttachmentDialog(new Uri($"{httpContextAccessor.HttpContext.Request.Scheme}://{httpContextAccessor.HttpContext.Request.Host.Value}")));    
            AddDialog(new WaitForProactiveDialog(httpContextAccessor, continuationParametersStore));
            AddDialog(new AuthDialog(configuration));
            AddDialog(new SsoSkillDialog(configuration));
            AddDialog(new FileUploadDialog());
            AddDialog(new DeleteDialog());
            AddDialog(new UpdateDialog());

            AddDialog(CreateEchoSkillDialog(conversationState, conversationIdFactory, skillClient, configuration));

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { ProcessActivityAsync }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        private static SkillDialog CreateEchoSkillDialog(ConversationState conversationState, SkillConversationIdFactoryBase conversationIdFactory, SkillHttpClient skillClient, IConfiguration configuration)
        {
            var botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;

            var skillHostEndpoint = configuration.GetSection("SkillHostEndpoint")?.Value;
            if (string.IsNullOrWhiteSpace(skillHostEndpoint))
            {
                throw new ArgumentException("SkillHostEndpoint is not in configuration");
            }

            var skillInfo = configuration.GetSection("EchoSkillInfo").Get<BotFrameworkSkill>() ?? throw new ArgumentException("EchoSkillInfo is not set in configuration");

            var skillDialogOptions = new SkillDialogOptions
            {
                BotId = botId,
                ConversationIdFactory = conversationIdFactory,
                SkillClient = skillClient,
                SkillHostEndpoint = new Uri(skillHostEndpoint),
                ConversationState = conversationState,
                Skill = skillInfo
            };
            var echoSkillDialog = new SkillDialog(skillDialogOptions);

            echoSkillDialog.Id = _echoSkill;
            return echoSkillDialog;
        }

        private async Task<DialogTurnResult> ProcessActivityAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // A skill can send trace activities, if needed.
            await stepContext.Context.TraceActivityAsync($"{GetType().Name}.ProcessActivityAsync()", label: $"Got ActivityType: {stepContext.Context.Activity.Type}", cancellationToken: cancellationToken);

            switch (stepContext.Context.Activity.Type)
            {
                case ActivityTypes.Event:
                    return await OnEventActivityAsync(stepContext, cancellationToken);

                default:
                    // We didn't get an activity type we can handle.
                    await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Unrecognized ActivityType: \"{stepContext.Context.Activity.Type}\".", inputHint: InputHints.IgnoringInput), cancellationToken);
                    return new DialogTurnResult(DialogTurnStatus.Complete);
            }
        }

        // This method performs different tasks based on the event name.
        private async Task<DialogTurnResult> OnEventActivityAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var activity = stepContext.Context.Activity;
            await stepContext.Context.TraceActivityAsync($"{GetType().Name}.OnEventActivityAsync()", label: $"Name: {activity.Name}. Value: {GetObjectAsJsonString(activity.Value)}", cancellationToken: cancellationToken);

            // Resolve what to execute based on the event name.
            switch (activity.Name)
            {
                case "Cards":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(CardDialog)).Id, cancellationToken: cancellationToken);

                case "Proactive":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(WaitForProactiveDialog)).Id, cancellationToken: cancellationToken);

                case "MessageWithAttachment":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(MessageWithAttachmentDialog)).Id, cancellationToken: cancellationToken);

                case "Auth":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(AuthDialog)).Id, cancellationToken: cancellationToken);

                case "Sso":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(SsoSkillDialog)).Id, cancellationToken: cancellationToken);

                case "FileUpload":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(FileUploadDialog)).Id, cancellationToken: cancellationToken);
                
                case "Echo":
                    // Start the EchoSkillBot
                    var messageActivity = MessageFactory.Text("I'm the echo skill bot");
                    messageActivity.DeliveryMode = stepContext.Context.Activity.DeliveryMode;
                    return await stepContext.BeginDialogAsync(FindDialog(_echoSkill).Id, new BeginSkillDialogOptions { Activity = messageActivity }, cancellationToken);

                case "Delete":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(DeleteDialog)).Id, cancellationToken: cancellationToken);

                case "Update":
                    return await stepContext.BeginDialogAsync(FindDialog(nameof(UpdateDialog)).Id, cancellationToken: cancellationToken);

                default:
                    // We didn't get an event name we can handle.
                    await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Unrecognized EventName: \"{activity.Name}\".", inputHint: InputHints.IgnoringInput), cancellationToken);
                    return new DialogTurnResult(DialogTurnStatus.Complete);
            }
        }

        private string GetObjectAsJsonString(object value) => value == null ? string.Empty : JsonConvert.SerializeObject(value);
    }
}
